bokeh/protocol.py

```diff
@@ -108,12 +137,15 @@ def default(self, obj):
         from .plot_object import PlotObject
         from .properties import HasProps
         from .colors import Color
+        from .models import ColumnDataSource
         ## array types
         if is_pandas and isinstance(obj, (pd.Series, pd.Index)):
             return self.transform_series(obj)
         elif isinstance(obj, np.ndarray):
             return self.transform_array(obj)
         elif isinstance(obj, PlotObject):
+            if isinstance(obj, ColumnDataSource):
+                obj.data.update(self.transform_column_source_data(obj.data))
```

Comment:
Right now `transform_column_data_source` seems like it is repeating the code in `default`. So I just had a new idea that might simplify this (probably should have occurred to me earlier). What about:
```
obj.data.update(BokehJSONEncoder.default(self, obj.data)
```
And then what you originally tried, putting an `isinstance(obj, (list, tuple))` check  *here*, that calls out to a `transform_sequence` should work, I think. Then this same machinery will handle any pandas series or numpy arrays inside `obj.data`. Worth a try I think.